### PR TITLE
MULE-15500: PollingSources don't release connections when item is rejected

### PR DESCRIPTION
**Pull request by ndinu**

### CODE REVIEW

File: modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/source/PollingSourceTestCase.java [hunk 1]
```diff
@@ -27,6 +26,8 @@
 import java.util.List;
 
 import org.junit.Test;
+import org.mule.test.petstore.extension.PetStoreConnectionProvider;
```
Comment (estebanwasinger) [inline modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/source/PollingSourceTestCase.java]: Order imports based on https://docs.google.com/document/d/1N2Axq3ABJjYmj9BjfYQkvJfF0odHQnlb8MAdVQydOWI/edit

File: modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/source/PollingSourceTestCase.java [hunk 2]
```diff
@@ -115,6 +122,15 @@ public void multiplePhasesOfWatermarkWithIncreasingAndDecreasingWatermarksPoll()
     assertAllNumbersAdoptedExactlyOnce();
   }
 
+  @Test
+  public void watermarkedItemsReleaseConnectionsTestCase() throws Exception {
+    startFlow("watermarkWithPooledConnection");
+    List<PetStoreConnectionProvider> petStoreConnectionProviders = PetStoreConnectionProvider.connections.keySet().stream()
+        .filter(x -> x instanceof PooledPetStoreConnectionProvider).collect(toList());
+    assertThat(petStoreConnectionProviders, hasSize(1));
```
Comment (alegmarra) [inline modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/source/PollingSourceTestCase.java]: polling prober

File: tests/test-extensions/petstore-extension/src/main/java/org/mule/test/petstore/extension/PetStoreConnectionProvider.java [hunk 3]
```diff
@@ -59,10 +62,14 @@
   @Parameter
   protected List<LocalDateTime> discountDates;
 
-  private int initialise, start, stop, dispose = 0;
+  private int initialise, start, stop, dispose, timesConnected = 0;
+
+  public static Map<PetStoreConnectionProvider, Integer> connections = new HashMap<>();
```
Comment (alegmarra) [inline tests/test-extensions/petstore-extension/src/main/java/org/mule/test/petstore/extension/PetStoreConnectionProvider.java]: why doing all this logic of handling the connections per provider type?

File: modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/source/PollingSourceReleasesConnectionTestCase.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+import org.junit.Test;
+
+public class PollingSourceReleasesConnectionTestCase extends AbstractExtensionFunctionalTestCase {
+
+  @Override
+  protected String getConfigFile() {
+    return "polling-source-releases-connection-config.xml";
+  }
+
```
Comment (alegmarra) [inline modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/source/PollingSourceReleasesConnectionTestCase.java]:
```
  @Override
  protected void doSetUpBeforeMuleContextCreation() throws Exception {
    PooledPetStoreConnectionProvider.TIMES_CONNECTED = 0;
}
```

**Pull request merged**